### PR TITLE
Healthcheck

### PR DESCRIPTION
**Pull request by nemchik**
Per discussion on Discord with @thelamer these changes should be easy to add to any LSIO image.

The `root/healthcheck.sh` file has two important lines in it:

- to check for the process that should be running
- to check that the nginx config tests valid

The process check (`pgrep`) should work for any other app as well. For example Sonarr would be `pgrep -af NzbDrone`.

The second check for Sonarr would be something more like `curl -f "http://localhost:8989/" >/dev/null`.

The reason `curl` is not used in this instance is http is not enabled by default in this container and `curl` will return an error looking for https because the cert would not match localhost. This could be changed to use the `URL` environment variable and `curl` could be used for this container as well if that is preferred. There is also no reason why there couldn't be more than two checks, so the `curl` check could be included in addition to the nginx config check.

A minimum of two checks per container is probably a fair starting point (process is running and web interface is accessible in most cases - those without a web interface would need a different type of second check). A maximum number checks would likely also be a good idea to define (personally I think anything more than 4 is excessive as these will run on an interval).

Each app may also require slight adjustments to the `interval`, `timeout`, `start-period`, or `retries`. These are defined explicitly in case the default values change. According to https://github.com/rcjsuen/dockerfile-utils/blob/b2df935062b66ec2b393a974414e73da3002de01/VALIDATOR.md#healthcheck "
Interval start-period cannot be less than 1ms" however the official documentation lists the default value to be `0s`.

Link for reference:
https://docs.docker.com/engine/reference/builder/#healthcheck

### CODE REVIEW

**Comment by nemchik** [comment]
Also for reference Plex has this in their official image, which would be an ideal second check for LSIO's Plex image as well:
https://github.com/plexinc/pms-docker/blob/28699d2a36948df108b8b27278b756d3eb2a07dc/Dockerfile#L57
https://github.com/plexinc/pms-docker/blob/28699d2a36948df108b8b27278b756d3eb2a07dc/root/healthcheck.sh

I chose to begin with this image because I already had the repo forked.

**Comment by nemchik** [comment]
After reviewing the @LinuxServer-CI link I had the idea that the health check could be used to confirm a passing build once health checks are included in all images. There may be more to that than what I'm thinking though.

File: Dockerfile [hunk 1]
```diff
@@ -109,3 +109,8 @@ RUN \
 
 # add local files
 COPY root/ /
+
+# healthcheck
+RUN chmod +x /healthcheck.sh
```
Comment (thelamer) [inline Dockerfile]: This chmod is not needed if the file is already +x in the git repo

File: root/healthcheck.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/with-contenv bash
+set -euo pipefail
+IFS=$'\n\t'
+
+# universal check that the process is running
+pgrep -af nginx > /dev/null
```
Comment (thelamer) [inline root/healthcheck.sh]: This is not really infallible due to how s6 overlay works. it restarts on a loop so you have a race condition with the pgrep and us trying to start nginx over and over again if it is busted

File: root/healthcheck.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+# individualized check(s) specific to the app
+nginx -tq -c /config/nginx/nginx.conf > /dev/null
```
Comment (thelamer) [inline root/healthcheck.sh]: From a functionality standpoint , testing the config should be done in the init of the container and message the user and maybe block us from running Nginx, the test should be something functional for this specific container.

**Comment by thelamer** [comment]
I cannot tell you what a good is letsencrypt is healthy check would be, but this does not seem like it is it. 
Hopefully some main devs of this container can chime in on this thread.

File: root/healthcheck.sh (hunk 2 above)
Comment (nemchik) [reply root/healthcheck.sh]: To my knowledge the health check (even if it's reporting unhealthy) shouldn't be trying to restart the container at all. There are ways to do this, such as an external container that monitors the health status of other containers and restarts them if they are not healthy, but the health check in the container itself shouldn't be doing that.

I may be missing something but that's how I read the description in the documentation. From my experience running the official Plex image which has a health check it does not automatically restart when the status is not healthy.

File: root/healthcheck.sh (hunk 3 above)
Comment (nemchik) [reply root/healthcheck.sh]: I can agree with that. Like you said in your other comment I'm also not sure what would be a good test for this specific container. I mentioned testing the using the URL environment variable which I think would look something like this: `curl -f "https://${URL}/" >/dev/null`. I'm in agreement that input from the other devs is needed on this though.

File: Dockerfile (hunk 1 above)
Comment (nemchik) [reply Dockerfile]: This is true. I saw other `chmod` happening in the Dockerfile and thought that might have been preferred, but I'm happy to change the permissions on the file instead. My only comment against that is that file permissions in GitHub can be easy to overlook (in my opinion).

File: root/healthcheck.sh (hunk 2 above)
Comment (thelamer) [reply root/healthcheck.sh]: The health check is not doing anything, what I am saying is the health check has a very high likelyhood to return healthy even if the service is busted. S6 keeps starting the pid over an over again, so you have millisecond windows where pgrep is gonna return an OK in those start commands.

**Comment by nemchik** [comment]
Letsencrypt may be one of the less obvious containers to implement health checks for, and thus may end up being somewhat more generic if there's not a better way to do it. As I noted in the first post, apps like Sonarr should be more straightforward.

File: Dockerfile (hunk 1 above)
Comment (thelamer) [reply Dockerfile]: not for us

File: root/healthcheck.sh (hunk 3 above)
Comment (thelamer) [reply root/healthcheck.sh]: Why do a full curl? , wouldn't a 200 response code lookup be sufficient and leaner :
```
curl -ILs -w "%{http_code}" -o /dev/null google.com
```

Comment (nemchik) [reply root/healthcheck.sh]: Indeed, that should work, good idea!

File: root/healthcheck.sh (hunk 2 above)
Comment (nemchik) [reply root/healthcheck.sh]: Makes sense, and after testing with just this one I can see that you are correct. Would other containers all be the same? Or is this specific to nginx in some way? (I don't want to assume)

If checking that the process is running isn't going to help then maybe only one check per app makes sense.

**Comment by sparklyballs** [comment]
i'm not sure there's any merit at all with a healthcheck considering the applications in our containers are being monitored by the s6 overlay.

**Comment by thelamer** [comment]
It does very little , but on the other side of that is the layer is small and the checks should be designed in a way to be extremely low resource usage and zero codebase maintenance . 

The checks are for external programs like swarm , or portainer . 

In general though and in practice the deeper you dive into the subject matter the more you see how little traction this has gained . Even official docker images don't have them and kubernetes has not adopted this as they have their own readiness checks. 

All of this exists because it was a "show me what you got" from a community user , at no fault of @nemchik though I have to say I personally am not very impressed by the technology or Dockers method of implementation. Seems more like something they tacked on to keep up with Google.

**Comment by nemchik** [comment]
I believe the intention of healthchecks is to confirm the app inside the container is actually running and doing what it's supposed to be doing. Like @thelamer said I also believe it's mostly beneficial to external apps like portainer or something that would automatically take action when a container is not healthy.

I don't know much about s6 so if it's doing this job already then I would agree this doesn't help much. But just to be clear the goal is to be sure the app on the inside is running properly, not to confirm that the container is running, and not to simply restart when there's a problem (although I mentioned it above, restarting isn't the only thing people might do with the health status, remediation efforts can happen too).

**Comment by thelamer** [comment]
@nemchik I really do appreciate the effort on this, but I am killing this PR. 

The further and further you read into health check the less and less appealing it looks. 

I will tell you what, as soon as this stuff is being implemented on official docker images here https://github.com/docker-library we can circle back to this and start an implementation plan. As I see it right now I do not believe Docker will support this going forward as they have made no honest effort to implement it on any of their base images. 
Seems like a left hand vs the right hand type of thing in their Org.

**Comment by nemchik** [comment]
Fair enough, I appreciate the time your team has spent reviewing this idea. I see their validity and hope that Docker does more with them in the future. In the meantime healthchecks can be added in compose (and probably via run/create commands as well) by end users who are determined to have them.

**Pull request closed without merging**